minetest/minetest
# Visual Effects Vol. 1

This PR is based on #14508 and adds a variety of effects to enhance the visual experience. To this end, it makes changes to shader code, as well as the engine source code, the Lua API and settings.

<details>
<summary>List of Features and Changes</summary>

- "soft" clouds look
- Tinted shadows
- Crude water reflections (sky and sun) and waves
- Translucent foliage
- Node specular highlights
- Adjusted fog color (more saturated where the fog is lighter)
- Minor changes to volumetric lighting (crudely simulates the effect of depth)

### Lua API additions
- Cloud parameters: `shadow` field controls the color of the cloud's base
- Lighting: `tint` controls shadow color
    - default would be black `{ r = 0, g = 0, b = 0}`
    - ambient light from the sky could be emulated by a blue color like `{ r = 0, g = 50, b = 180 }`

### Settings
| Name | Technical Name | Location | Type | Default |
|---|---|---|---|---|
| Soft clouds | soft_clouds | Graphics>Clouds | bool | false |
| Liquid reflections | enable_water_reflections | Shaders>Waving Nodes | bool | false |
| Translucent Foliage | enable_translucent_foliage | Shaders>Other Effects | bool | false |
| Node specular | enable_node_specular | Shaders>Other Effects | bool | false |
</details>

<details>
<summary>Screenshots</summary>

![screenshot_20240504_201730](https://github.com/minetest/minetest/assets/72752000/78a70705-b3a5-4077-adc3-133d5d18768e)
![screenshot_20240504_201612](https://github.com/minetest/minetest/assets/72752000/37f1684b-522c-4a7d-9496-098e5062554a)
![screenshot_20240504_201455](https://github.com/minetest/minetest/assets/72752000/74c167e5-cae0-4082-9363-c6191a7fd191)
![screenshot_20240504_202011](https://github.com/minetest/minetest/assets/72752000/afdadcb8-2001-46cd-bb2f-43f81923b009)

</details>

## To do

This PR is Ready for Review.

## How to test

For most features, simply change the game settings. Of course, the game you use should support dynamic shadows and volumetrics.
The added Lua properties can be used like so:
```Lua
player:set_clouds({shadow = { r = 120, g = 160, b = 240 }})
player:set_lighting(shadows = { intensity = 0.5, tint = { r = 0, g = 50, b = 180 }})
```
